cedar-policy-core/src/parser/err.rs

```diff
@@ -47,9 +54,288 @@ pub enum ParseError {
     /// Error in the CST -> AST transform, mostly well-formedness issues.
     #[error("poorly formed: {0}")]
     #[diagnostic(code(cedar_policy_core::parser::to_ast_err))]
-    ToAST(String),
+    ToAST(#[from] ToASTError),
+    /// Error concerning restricted expressions.
+    #[error(transparent)]
+    RestrictedExpressionError(#[from] RestrictedExprError),
+    /// Errors concerning parsing literals on their own
+    #[error("{0}")]
+    ParseLiteral(#[from] ParseLiteralError),
+}
+
+/// Errors in the top-level parse literal entrypoint
+#[derive(Debug, Clone, PartialEq, Error, Eq)]
+pub enum ParseLiteralError {
+    /// The top-level parser endpoint for parsing a literal encountered a non-literal.
+    /// Since this can be any possible other expression, we just return it as a string.
+    #[error("the source `{0}` is not a literal")]
+    ParseLiteral(String),
+}
+
+/// Errors in  the CST -> AST transform, mostly well-formedness issues.
+#[derive(Debug, Diagnostic, Error, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ToASTError {
+    /// Returned when we attempt to parse a template with a conflicting id
+    #[error("a template with this id already exists in the policy set: {0}")]
+    DuplicateTemplateID(PolicyID),
+    /// Returned when we attempt to parse a policy with a conflicting ID
+    #[error("a policy with this ID already exists in the policy set: {0}")]
+    DuplicatePolicyID(PolicyID),
+    /// Returned when we attempt to parse a policy with malformed or conflicting annotations
+    #[error("this policy uses poorly formed or duplicate annotations")]
+    BadAnnotations,
+    /// Returned when a policy contains Template Slots in the condition clause. This is not currently supported.
+    #[error("template slots are currently unsupported in policy condition clauses")]
+    SlotsInConditionClause,
+    /// Returned when a policy is missing one of the 3 required scope clauses. (`principal`, `action`, and `resource`)
+    #[error("this policy is missing the {0} variable in the scope")]
+    MissingScopeConstraint(Var),
+    /// Returned when a policy has an extra scope clause. This is not valid syntax
+    #[error("this policy has extra variable in the scope A policy must have exactly `principal`, `action`, and `resource`")]
+    ExtraHeadConstraints,
```

Comment:
It would be helpful for this error to carry text representing what the extra variable was